pallets/xcmp-queue/src/lib.rs

```diff
@@ -941,84 +510,160 @@ impl<T: Config> Pallet<T> {
+			.try_consume(T::WeightInfo::enqueue_xcmp_messages(xcms.len() as u32))
+			.is_err()
+		{
+			return
+		}
+		let QueueConfigData { drop_threshold, .. } = <QueueConfig<T>>::get();
+		let fp = T::XcmpQueue::footprint(sender);
+
+		let new_count = xcms.len().saturating_add(fp.count as usize);
+		let to_enqueue = (drop_threshold as usize).saturating_sub(new_count) as usize;
+		if to_enqueue < xcms.len() {
+			// This should not happen since the channel should have been suspended in [`on_queue_changed`].
+			log::error!(
+				"XCMP queue for sibling {:?} is full; dropping {} messages.",
+				sender,
+				xcms.len() - to_enqueue,
+			);
+			xcms.truncate(to_enqueue);
+		}
+
+		T::XcmpQueue::enqueue_messages(xcms.iter().map(|xcm| xcm.as_bounded_slice()), sender);
+	}
+
+	/// Split concatenated encoded `VersionedXcm`s into individual items.
+	///
+	/// We directly encode them again since that is needed later on.
+	fn split_concatenated_xcms(
+		data: &mut &[u8],
+	) -> Result<Vec<BoundedVec<u8, MaxXcmpMessageLenOf<T>>>, ()> {
+		// FAIL-CI add benchmark to check for OOM depending on `MAX_XCM_DECODE_DEPTH`.
+		let mut encoded_xcms = Vec::new();
+		while !data.is_empty() {
+			let xcm =
+				VersionedXcm::<T::RuntimeCall>::decode_with_depth_limit(MAX_XCM_DECODE_DEPTH, data)
+					.map_err(|_| ())?;
+			let bounded = xcm.encode().try_into().map_err(|_| ())?;
+			encoded_xcms.push(bounded);
+		}
+		Ok(encoded_xcms)
+	}
+}
+
+impl<T: Config> OnQueueChanged<ParaId> for Pallet<T> {
+	// Suspends/Resumes the queue when certain thresholds are reached.
+	fn on_queue_changed(para: ParaId, count: u64, _size: u64) {
+		let QueueConfigData { resume_threshold, suspend_threshold, .. } = <QueueConfig<T>>::get();
+
+		let mut suspended_channels = <InboundXcmpSuspended<T>>::get();
+		let suspended = suspended_channels.contains(&para);
+
+		if suspended && count <= resume_threshold as u64 {
+			if let Err(err) = Self::send_signal(para, ChannelSignal::Resume) {
+				log::error!("Cannot resume channel from sibling {:?}: {:?}", para, err);
+			} else {
+				suspended_channels.remove(&para);
+				<InboundXcmpSuspended<T>>::put(suspended_channels);
+			}
+		} else if !suspended && count >= suspend_threshold as u64 {
+			log::warn!("XCMP queue for sibling {:?} is full; suspending channel.", para);
+
+			if let Err(err) = Self::send_signal(para, ChannelSignal::Suspend) {
+				// This is an edge-case, but we will not regard the channel as `Suspended` without
+				// confirmation. It will just re-try to suspend in the next block.
+				log::error!("Cannot suspend channel from sibling {:?}: {:?}; further messages may be dropped.", para, err);
+			} else if let Err(err) = suspended_channels.try_insert(para) {
+				log::error!("Too many channels suspended; cannot suspend sibling {:?}: {:?}; further messages may be dropped.", para, err);
+			} else {
+				<InboundXcmpSuspended<T>>::put(suspended_channels);
+			}
+		}
+	}
+}
+
+impl<T: Config> QueuePausedQuery<ParaId> for Pallet<T> {
```

Comment:
Maybe also worth changing to:
```rust
impl<T: Config> QueuePausedQuery<cumulus_primitives_core::AggregateMessageOrigin> for Pallet<T> {
```

Then we may remove the `NarrowToSiblings` and use the `XcmpQueue` directly:
```rust
type QueuePausedQuery = XcmpQueue;
```

Reply:
But then we have the problem that there is an "impossible" case in the code, where a non-sibling origin is passed to the function.  
To rule this out directly, i think its better to have an adapter á la `NarrowToSiblings`.

Reply:
Ok, this works too :)